packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/platform_webview_permission_request.dart

```diff
@@ -0,0 +1,162 @@
+  ///
+  /// This should only be used by this class and subclasses in platform
+  /// implementations.
+  @protected
+  const WebViewPermissionResourceType(this.name);
+
+  /// Unique name of the resource type.
+  ///
+  /// For platform implementations, this should match the name of variable.
+  final String name;
+
+  /// A media device that can capture video.
+  static const WebViewPermissionResourceType camera =
+      WebViewPermissionResourceType('camera');
+
+  /// A media device that can capture audio.
+  static const WebViewPermissionResourceType microphone =
+      WebViewPermissionResourceType('microphone');
+}
+
+/// Permissions request when web content requests access to protected resources.
+///
+/// A response MUST be provided by calling a method from [response].
+///
+/// Platform specific implementations can add additional methods when extending
+/// this class.
+///
+/// This example demonstrates how to extend the
+/// [PlatformWebViewPermissionRequest] to provide additional platform-specific
+/// features:
+///
+/// ```dart
+/// class WebKitWebViewPermissionRequest extends PlatformWebViewPermissionRequest {
+///   const WebKitWebViewPermissionRequest._({
+///     required super.types,
+///     required void Function(WKPermissionDecision decision) onDecision,
+///   }) : _onDecision = onDecision;
+///
+///   final void Function(WKPermissionDecision) _onDecision;
+///
+///   @override
+///   Future<void> grant(WebViewPermissionGrantParams params) async {
+///     _onDecision(WKPermissionDecision.grant);
+///   }
+///
+///   @override
+///   Future<void> deny(WebViewPermissionDenyParams params) async {
+///     _onDecision(WKPermissionDecision.deny);
+///   }
+///
+///   Future<void> prompt(WebViewPermissionDenyParams params) async {
+///     _onDecision(WKPermissionDecision.prompt);
+///   }
+/// }
+/// ```
+@immutable
+abstract class PlatformWebViewPermissionRequest {
+  /// Creates a [PlatformWebViewPermissionRequest].
+  const PlatformWebViewPermissionRequest({required this.types});
+
+  /// All resources access has been requested for.
+  final List<WebViewPermissionResourceType> types;
+
+  /// Grant permission for the requested resource(s).
+  Future<void> grant(WebViewPermissionGrantParams params);
+
+  /// Deny permission for the requested resource(s).
+  Future<void> deny(WebViewPermissionDenyParams params);
+}
+
+/// Params for granting permission for a requested resource.
+///
+/// Platform specific implementations can add additional fields by extending
+/// this class.
+///
+/// When extending [WebViewPermissionGrantParams] additional parameters should
+/// always accept `null` or have a default value to prevent breaking changes.
+///
+/// This example demonstrates how to extend the [WebViewPermissionGrantParams]
+/// to provide additional platform specific parameters:
+///
+/// ```dart
+/// class AndroidWebViewPermissionGrantParams extends WebViewPermissionGrantParams {
```

Comment:
Do we have a use case in mind for these param objects? Are there known examples of needing to provide more params for these responses?

(Since we have the escape hatch of platforms being able to add their own response methods anyway, I'm wondering if this is overkill.)

Reply:
If the WebView requests audio and camera, Android has the ability to only grant camera. But you're correct that this could be handled by a method, such as `grantOnly`.